File: Project_Prediction/Own_Running_Data_Analysis_and_Prediction.R

install.packages('pacman')

library(pacman)
pacman:: p_load(ggplot2,scales,dplyr,DT,sqldf,readxl,corrplot)

# Specify the file paths for the Apple and Garmin datasets
file_path_apple <- "C:/Users/fbeik/OneDrive/Desktop/Project/Activities_Apple.csv"
file_path_garmin <- "C:/Users/fbeik/OneDrive/Desktop/Project/Activities_Garmin.csv"

# Load the datasets from the specified file paths
data_apple <- read.csv(file_path_apple,  sep = ";", fill=TRUE)
data_garmin <- read.csv(file_path_garmin,  sep = ";", fill=TRUE)


# Exploratory Data Analysis
print(head(data_apple))
print(head(data_garmin))

data_apple <- sqldf (" SELECT `ï..ID` AS ID, Date, `Part.Of.Day` AS PartOfDay, `Activity.Type` AS ActivityType, Distance, `Elapsed.Time` AS TimeSeconds, `Max.Heart.Rate` AS MaxHeartRate, round(`Average.Heart.Rate`) AS AverageHeartRate, Calories,round(`Elevation.Gain`,2) AS ElevationGain, round(`Elevation.Loss`,2) AS ElevationLoss, `Relative.Effort` AS RelativeEffort FROM data_apple")
data_garmin <- sqldf (" SELECT `ï..ID` AS ID, Date, `Time.of.day` AS PartOfDay, `Type` AS ActivityType, Distance, `Time` AS TimeSeconds, `Max.Heart.Rate` AS MaxHeartRate, round(`Avg..Heart.Rate`) AS AverageHeartRate, Calories,round(`Elevation.Gain`,2) AS ElevationGain, round(`Elevation.Loss`,2) AS ElevationLoss, `Relative.Effort` AS RelativeEffort FROM data_garmin")

# Combine the Apple and Garmin datasets into a single dataset
data <- sqldf("
    SELECT * 
    FROM data_apple
    UNION ALL
    SELECT *
    FROM data_garmin
")

# Exploratory Data Analysis
print(head(data))
summary(data)

# Check for duplicates and get the total number of duplicates
sum(duplicated(data))
# View the duplicated rows
data[duplicated(data), ]
# removing duplicate data aand check how many duplicates were removed
duplicates_before <- nrow(data)
data <- unique(data)
duplicates_after <- nrow(data)
cat("Duplicates removed:", duplicates_before - duplicates_after, "\n")


# Remove rows where the 'ID' column is NA
data <- data %>% filter(!is.na(ID))

# check the datatypes
str(data)

### Calculation for Pacer per KM
# Add a new column to convert time from seconds to minutes
data$time_in_minutes <- data$TimeSeconds / 60
# Replace commas with periods in the 'Distance' column and make it numeric
data$Distance <- as.numeric(gsub(",", ".", data$Distance))
# Calculate the pace (minutes per kilometer)
data$PacePerKM <- data$time_in_minutes / as.numeric(data$Distance)
# Convert the calculated pace to a time format (MM:SS)
data$PacePerKM <- sapply(data$PacePerKM, function(pace) {
  minutes <- as.integer(floor(pace))  # Extract whole minutes
  seconds <- as.integer(round((pace - minutes) * 60))  # Calculate remaining seconds
  sprintf("%02d:%02d", minutes, seconds)  # Format as MM:SS
})


# # Ensure the 'Time' column is numeric and convert it to "HH:MM:SS" format
data$TimeSeconds <- as.numeric(data$TimeSeconds)
data <- data[!is.na(data$TimeSeconds), ]  # Remove rows with NA
data$Time1 <- format(as.POSIXct(data$TimeSeconds, origin = "1970-01-01", tz = "UTC"), "%H:%M:%S")

# Filter for only running activities and exclude runs under 1 km; sort by distance
data_run <- sqldf("SELECT ID, Date, PartOfDay, ActivityType, Distance, TimeSeconds,`Time1` AS Time, PacePerKM, MaxHeartRate, AverageHeartRate, Calories, ElevationGain, ElevationLoss, RelativeEffort  FROM data")

data_run <- data %>%
  filter(ActivityType == "Run", Distance > 0.99) %>%
  arrange(desc(Distance))

# Convert the 'Date' column to a new format and add the weekday information
Sys.setlocale("LC_TIME", "C")
data_run$FormattedDate <- format(strptime(data_run$Date, format = "%b %d, %Y, %I:%M:%S %p"),"%Y-%m-%d %H:%M:%S")
WeekdayDate <- format(strptime(data_run$FormattedDate, format = "%Y-%m-%d %H:%M:%S"), "%m/%d/%Y")
data_run$Weekday <- weekdays(WeekdayDate)

# Correlation matrix to understand relationships between numerical features
cor_matrix <- cor(data_run[, sapply(data_run, is.numeric)])
print(cor_matrix)
# Visualize correlation matrix
corrplot(cor_matrix, method = "circle",type = "lower")


# Analyze running activity by weekday: count runs and calculate the average distance
weekday_summary <- data_run %>%
  group_by(Weekday) %>%
  summarise(Runs = n(),
            Avg_Distance = mean(Distance)) %>%
  arrange(desc(Runs))

print(weekday_summary)

# Analyze running activity by time of day: count runs and calculate the average distance
time_summary <- data_run %>%
  group_by(PartOfDay) %>%
  summarise(
    Runs = n(),
    Avg_Distance = mean(Distance)) %>%
  arrange(desc(Runs))

print(time_summary)

# Analyze running activity by both weekday and time of day
all_summary <- data_run %>%
  group_by(Weekday, PartOfDay) %>%
  summarise(
    Runs = n(),
    Avg_Distance = mean(Distance)) %>%
  arrange(Weekday,desc(Runs))

print(all_summary, n=27)

# Convert 'PartOfDay' into a factor with a specific order for visualization
all_summary$PartOfDay <- factor(all_summary$PartOfDay, levels = c("Morning", "Lunch", "Afternoon", "Evening", "Night"),exclude = NULL)

# Visualize the number of runs by weekday and time of day
ggplot(data = all_summary, aes(x = Weekday, y = Runs, fill = PartOfDay)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "Läufe nach Wochentag und Tageszeit", y = "Anzahl der Läufe", x = "Wochentag") +
  theme_minimal()



# Count runs with and without elevation change, indicating whether they were indoors or outdoors
no_elevation_change <- nrow(data_run[data_run$ElevationGain == 0.0 & data_run$ElevationLoss == 0.0, ])
with_elevation_change <- nrow(data_run[data_run$ElevationGain > 0 | data_run$ElevationLoss > 0, ])
# Add a column to the table to show if the run was inside or outside
data_run$RunCondition <- ifelse(data_run$ElevationGain == 0 & data_run$ElevationLoss == 0, 
  "inside", 
  "outside"
)

cat("Runs without elevation change (inside):", no_elevation_change, "\n")
cat("Runs with elevation change (outside):", with_elevation_change, "\n")

# Calculate percentage of inside and outside runs
total_runs <- nrow(data_run)
percentage_inside <- (no_elevation_change / total_runs) * 100
percentage_outside <- (with_elevation_change / total_runs) * 100

cat("Percentage of runs inside:", round(percentage_inside, 2), "%\n")
cat("Percentage of runs outside:", round(percentage_outside, 2), "%\n")



# Calculate the average distance and the total distance of all runs
average_distance <- mean(data_run$Distance)
total_distance  <- sum(data_run$Distance)

# Add split times data from another file and join it with the running data
file_path_splits <- "C:/Users/fbeik/OneDrive/Desktop/Project/Activities_splits.csv"
data_splits <- read.csv(file_path_splits,  sep = ";", fill=TRUE)
data_splits <- sqldf("SELECT `ï..ID` AS ID, Split1, Split2, Split3, Split4, Split5, Split6, Split7, Split8, Split9, Split10, Split11, Split12, Split13, Split14, Split15, Split16, Split17, Split18, Split19, Split20, Split21, Split22 FROM data_splits")
joined_data <- left_join(data_run, data_splits, by = "ID")

joined_data$Distance <- as.numeric(joined_data$Distance)
# Visualize the relationship between distance and time to determine if it is linear or non-linear
ggplot(joined_data, aes(x = Distance, y = Time)) +
  geom_point() +
  geom_smooth(method = "lm", color = "green") +
  labs(title = "Relationship Between Distance and Time")

# Predict marathon time using Riegel's formula (non-linear extrapolation)
riegel_predict <- function(TimeShortDistance, ShortDistance, LongDistance, exponent = 1.06) {
  TimeLongDistance <- TimeShortDistance * (LongDistance / ShortDistance)^exponent
  return(TimeLongDistance)
}


# Predict marathon time based on average runs between 11 and 21 km
TimeShortDistance <- mean(joined_data$TimeSeconds[joined_data$Distance > 11]) 
ShortDistance <- mean(joined_data$Distance[joined_data$Distance > 11])
LongDistance <- 42.195
predicted_time_riegel <- riegel_predict(TimeShortDistance, ShortDistance, LongDistance)
Predicted_Marathon_Time <- format(as.POSIXct(predicted_time_riegel, origin = "1970-01-01", tz = "UTC"), "%H:%M:%S")
cat("Predicted Marathon Time (seconds):", Predicted_Marathon_Time)

File: Project_Prediction/Own_Running_Data_Analysis_and_Prediction.py
import pandas as pd
import numpy as np
import math
from dateutil import parser
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import GridSearchCV
import xgboost as xgb

# Specifying the file paths for Apple and Garmin datasets
file_path_apple = "C:/Users/fbeik/OneDrive/Desktop/Project/Activities_Apple_FB.csv"
file_path_garmin = "C:/Users/fbeik/OneDrive/Desktop/Project/Activities_Garmin_FB.csv"

# Loading the datasets from the specified file paths
data_apple = pd.read_csv(file_path_apple, sep=";")
data_garmin = pd.read_csv(file_path_garmin, sep=";")

# Exploratory Data Analysis: Printing the first few rows
print("Apple Data (first 5 rows):")
print(data_apple.head())
print("\nGarmin Data (first 5 rows):")
print(data_garmin.head())

# Renaming and transforming columns for the Apple dataset
data_apple = data_apple.rename(columns={
    "ï..ID": "ID",
    "Part Of Day": "PartOfDay",
    "Activity Type": "ActivityType",
    "Elapsed Time": "TimeSeconds",
    "Max Heart Rate": "MaxHeartRate",
    "Average Heart Rate": "AverageHeartRate",
    "Elevation Gain": "ElevationGain",
    "Elevation Loss": "ElevationLoss",
    "Relative Effort": "RelativeEffort"
})

# Ensuring the columns are numeric and round
data_apple['AverageHeartRate'] = pd.to_numeric(data_apple['AverageHeartRate'], errors='coerce')
data_apple['ElevationGain'] = pd.to_numeric(data_apple['ElevationGain'], errors='coerce')
data_apple['ElevationLoss'] = pd.to_numeric(data_apple['ElevationLoss'], errors='coerce')

data_apple['AverageHeartRate'] = data_apple['AverageHeartRate'].round(0)
data_apple['ElevationGain'] = data_apple['ElevationGain'].round(2)
data_apple['ElevationLoss'] = data_apple['ElevationLoss'].round(2)

# Renaming and transforming columns for the Garmin dataset
data_garmin = data_garmin.rename(columns={
    "ï..ID": "ID",
    "Time.of.day": "PartOfDay",
    "Type": "ActivityType",
    "Time": "TimeSeconds",
    "Max Heart Rate": "MaxHeartRate",
    "Avg. Heart Rate": "AverageHeartRate",
    "Elevation Gain": "ElevationGain",
    "Elevation Loss": "ElevationLoss",
    "Relative Effort": "RelativeEffort"
})

# Ensuring the columns are numeric and round
data_garmin['AverageHeartRate'] = pd.to_numeric(data_garmin['AverageHeartRate'], errors='coerce')
data_garmin['ElevationGain'] = pd.to_numeric(data_garmin['ElevationGain'], errors='coerce')
data_garmin['ElevationLoss'] = pd.to_numeric(data_garmin['ElevationLoss'], errors='coerce')

data_garmin['AverageHeartRate'] = data_garmin['AverageHeartRate'].round(0)
data_garmin['ElevationGain'] = data_garmin['ElevationGain'].round(2)
data_garmin['ElevationLoss'] = data_garmin['ElevationLoss'].round(2)


# Combining the Apple and Garmin datasets into a single dataset
data = pd.concat([data_apple, data_garmin], ignore_index=True)

# Exploratory Data Analysis: Printing the first few rows and a summary of the combined data
print("\nCombined Data (first 5 rows):")
print(data.head())
print("\nSummary of Combined Data:")
print(data.describe(include='all'))


# Checking for duplicates and get the total number of duplicates
duplicates_count = data.duplicated().sum()
print("\nTotal number of duplicate rows:", duplicates_count)

# Viewing the duplicated rows
duplicated_rows = data[data.duplicated()]
print("\nDuplicated rows:")
print(duplicated_rows)

# Removing duplicate rows and check how many were removed
duplicates_before = data.shape[0]
data = data.drop_duplicates()
duplicates_after = data.shape[0]
print("\nDuplicates removed:", duplicates_before - duplicates_after)

# removing rows where ID is NA
data = data.dropna(subset=['ID'])

# checking datatypes
print(data.info())

# Calculating pace per km and adding time in minutes
data['time_in_minutes'] = pd.to_numeric(data['TimeSeconds'], errors='coerce') / 60

# Replacing commas with periods in the 'Distance' column and convert to numeric
data['Distance'] = data['Distance'].astype(str).str.replace(',', '.')
data['Distance'] = pd.to_numeric(data['Distance'], errors='coerce')

# Calculating the pace (minutes per kilometer)
data['PacePerKM_value'] = data['time_in_minutes'] / data['Distance']

# Converting the calculated pace to a time format (MM:SS)
def convert_pace_to_mmss(pace):
    if pd.isna(pace) or pace <= 0 or np.isinf(pace):
        return None
    minutes = int(math.floor(pace))
    seconds = int(round((pace - minutes) * 60))
    if seconds == 60:
        minutes += 1
        seconds = 0
    return f"{minutes:02d}:{seconds:02d}"

data['PacePerKM'] = data['PacePerKM_value'].apply(convert_pace_to_mmss)

# ensuring the 'TimeSeconds' column is numeric and convert it to "HH:MM:SS" format
data['TimeSeconds'] = pd.to_numeric(data['TimeSeconds'], errors='coerce')
data = data.dropna(subset=['TimeSeconds'])
data['Time1'] = pd.to_datetime(data['TimeSeconds'], unit='s', origin='unix').dt.strftime('%H:%M:%S')

# Filtering for only running activities and exclude runs under 1 km
data_run = data[(data['ActivityType'] == "Run") & (data['Distance'] > 0.99)].copy()
data_run = data_run.sort_values(by='Distance', ascending=False)

def robust_date_parser(date_str):
    try:
        return pd.to_datetime(date_str, format='%b %d, %Y, %I:%M:%S %p')
    except Exception:
        pass 
    try:
        return parser.parse(date_str)
    except Exception:
        # If all parsing attempts fail, return a default date
        print(f"Warning: Could not parse date '{date_str}'. Using default date 1900-01-01.")
        return pd.Timestamp("1900-01-01")

# Applying the parser to create FormattedDate column
data_run['FormattedDate'] = data_run['Date'].apply(robust_date_parser)

# Optionally, convert to a specific string format
#data_run['FormattedDate'] = data_run['FormattedDate'].dt.strftime('%Y-%m-%d %H:%M:%S')

# checking again for null values
null_count = data_run['FormattedDate'].isnull().sum()
print("Number of null values in 'FormattedDate':", null_count)

# Creating the Weekday column from FormattedDate
data_run['Weekday'] = pd.to_datetime(data_run['FormattedDate'], format='%Y-%m-%d %H:%M:%S').dt.day_name()

# Checking the results
print(data_run[['Date', 'FormattedDate', 'Weekday']].head())


numeric_data = data_run.select_dtypes(include=[np.number])

# Display the first few rows
print(numeric_data.head())

# Selecting only the specified numeric columns
selected_columns = ["Calories", "MaxHeartRate", "AverageHeartRate", "RelativeEffort", "TimeSeconds", "Distance", "time_in_minutes", "PacePerKM_value"]
    
# Creating a new DataFrame with the selected columns
numeric_data_selected = data_run[selected_columns]

# Selecting numeric columns and compute the correlation matrix
cor_matrix = numeric_data_selected.corr()
print("Correlation Matrix:")
print(cor_matrix)

# Visualizing the correlation matrix using a heatmap - displaying only the lower triangle
plt.figure(figsize=(10, 8))
mask = np.triu(np.ones_like(cor_matrix, dtype=bool))  # mask for the upper triangle
sns.heatmap(cor_matrix, mask=mask, annot=True, cmap='coolwarm', square=True)
plt.title("Correlation Matrix (Lower Triangle)")
plt.show()

# analyzing running activity by weekday
weekday_summary = (
    data_run.groupby("Weekday")
    .agg(Runs=("Weekday", "size"), Avg_Distance=("Distance", "mean"))
    .reset_index()
    .sort_values(by="Runs", ascending=False)
)
print("Weekday Summary:")
print(weekday_summary)

# analyzing running activity by time of day
time_summary = (
    data_run.groupby("PartOfDay")
    .agg(Runs=("PartOfDay", "size"), Avg_Distance=("Distance", "mean"))
    .reset_index()
    .sort_values(by="Runs", ascending=False)
)
print("Time of Day Summary:")
print(time_summary)

# analyzing running activity by both weekday and time of day
all_summary = (
    data_run.groupby(["Weekday", "PartOfDay"])
    .agg(Runs=("ID", "size"), Avg_Distance=("Distance", "mean"))
    .reset_index()
)
# Sorting by Weekday and descending by Runs
all_summary = all_summary.sort_values(by=["Weekday", "Runs"], ascending=[True, False])
print("All Summary:")
print(all_summary)

# Converting 'PartOfDay' to a categorical type with a specified order
categories = ["Morning", "Lunch", "Afternoon", "Evening", "Night"]
all_summary["PartOfDay"] = pd.Categorical(all_summary["PartOfDay"], categories=categories, ordered=True)

# visualizing the number of runs by weekday and time of day
plt.figure(figsize=(10, 6))
sns.barplot(data=all_summary, x="Weekday", y="Runs", hue="PartOfDay", dodge=True)
plt.title("Läufe nach Wochentag und Tageszeit")
plt.xlabel("Wochentag")
plt.ylabel("Anzahl der Läufe")
plt.show()

# calculating average and total distance of all runs
average_distance = data_run['Distance'].mean()
total_distance = data_run['Distance'].sum()
print("Average Distance (km):", average_distance)
print("Total Distance (km):", total_distance)

# visualizing relationship between distance and time
plt.figure(figsize=(8, 6))
sns.scatterplot(data=data_run, x="Distance", y="TimeSeconds")
sns.regplot(data=data_run, x="Distance", y="TimeSeconds", scatter=False, color="green")
plt.title("Relationship Between Distance and Time")
plt.show()

# Linear regression model
# Converting numerical columns with comma separators
for col in ["Distance", "MaxHeartRate", "AverageHeartRate", "ElevationGain", "Calories"]:
    data_run[col] = data_run[col].astype(str).str.replace(",", ".").astype(float)

print(data_run.isna().sum())  # Counting missing values in each column
# Filling missing values with mean of each column
data_run.fillna(data_run.mean(numeric_only=True), inplace=True)

# Selecting features and target variable
X = data_run[["Distance", "MaxHeartRate", "AverageHeartRate", "Calories"]]
y = data_run["TimeSeconds"]

# Train-Test Split (80%-20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Training Linear Regression Model
lm_model = LinearRegression()
lm_model.fit(X_train, y_train)

# Making predictions
y_pred = lm_model.predict(X_test)

# Evaluating model
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print("Linear Regression RMSE:", rmse)

# On average, the model's predictions differ by ~7 min 56 sec from the actual running time
# model is making large errors (~7 min 56 sec) - therefore testing another model

# Training Random Forest Model
rf_model = RandomForestRegressor(n_estimators=500, random_state=42)
rf_model.fit(X_train, y_train)

# Making predictions
y_pred_rf = rf_model.predict(X_test)

# Evaluating model
rmse_rf = np.sqrt(mean_squared_error(y_test, y_pred_rf))
print("Random Forest RMSE:", rmse_rf)

# Feature importance
feat_importances = pd.Series(rf_model.feature_importances_, index=X.columns)
feat_importances.sort_values().plot(kind="barh")
plt.title("Feature Importance in Random Forest")
plt.show()

# On average, the Random Forest model’s predictions are off by ~11 min 32 sec from the actual time

# XGBOOST
# Defining hyperparameter grid
param_grid = {
    "n_estimators": [100, 500, 1000],
    "max_depth": [3, 6, 10],
    "learning_rate": [0.01, 0.1, 0.2],
    "subsample": [0.7, 1.0]
}

# Grid Search for best parameters
xgb_model = xgb.XGBRegressor()
grid_search = GridSearchCV(xgb_model, param_grid, cv=5, scoring="neg_root_mean_squared_error")
grid_search.fit(X_train, y_train)

# Printing the best parameters
print("Best parameters:", grid_search.best_params_)

# Training model
xgb_best = xgb.XGBRegressor(**grid_search.best_params_)
xgb_best.fit(X_train, y_train)

# Predicting and calculating RMSE
y_pred_xgb_best = xgb_best.predict(X_test)
rmse_xgb_best = np.sqrt(mean_squared_error(y_test, y_pred_xgb_best))
print("Optimized XGBoost RMSE:", rmse_xgb_best)

# On average the XGBoost predictions are off by ~5 min 37 sec
# Outliers in Data – Large variation in running times can mislead the model
# Data might not be large enough – XGBoost works best with big datasets (thousands of rows).

# machine learning models struggle to generalize running performance
# Using Riegel’s formula - a well-established non-linear sports performance predictor
# Riegel's formula function
def riegel_predict(time_short, distance_short, distance_long, exponent=1.06):
    time_long = time_short * (distance_long / distance_short) ** exponent
    return time_long

# Filtering data for runs between 11km and 21km
filtered_runs = data_run[data_run["Distance"] > 11]

# Calculateing average running time for those runs
time_short_distance = filtered_runs["TimeSeconds"].mean()
short_distance = filtered_runs["Distance"].mean()

# Defining the target marathon distance
long_distance = 42.195  # Marathon distance in km

# Predicting marathon time
predicted_time_riegel = riegel_predict(time_short_distance, short_distance, long_distance)

# Converting seconds to HH:MM:SS format
predicted_marathon_time = pd.to_datetime(predicted_time_riegel, unit='s').strftime("%H:%M:%S")

print(f"Predicted Marathon Time: {predicted_marathon_time}")
